### ISSUE TITLE
failed iput logging as a failed mkdir

### ISSUE THREAD

**Issue by andyjp94** [irods/irods#3226]
So we were doing some stress testing on one of our dev systems with the following command:
```
for count in $(seq 1 1100); do time iput -K testfile /seq-dev/jc18_testcol/testfile_${count} | tee -a seq-dev_timing_log.20160714_1129.log; done
```
Some of these tests failed with the following error message:
```
Jul 14 11:43:20 pid:32087 ERROR: [-]    iRODS/server/core/src/fileOpr.cpp:250:mkFileDirR :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [fileMkdir for [/irods-seq-sr04-ddn-gc10-33-34-35/jc18_testcol]]
        [-]     iRODS/server/drivers/src/fileDriver.cpp:290:fileMkdir :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [failed to call 'mkdir']
                [-]     libpassthru.cpp:354:passthru_file_mkdir_plugin :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [passthru_file_mkdir_plugin - failed calling child mkdir.]
                        [-]     librepl.cpp:863:replFileMkdir :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [replFileMkdir - Failed while calling child operation.]
                                [-]     librandom.cpp:412:random_file_mkdir :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [Failed calling child operation.]
                                        [-]     librandom.cpp:412:random_file_mkdir :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [Failed calling child operation.]
                                                [-]     libunixfilesystem.cpp:978:unix_file_mkdir_plugin :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [Mkdir error for "/irods-seq-sr04-ddn-gc10-33-34-35/jc18_testcol", errno = "No space left on device", status = -520028.]

Jul 14 11:43:20 pid:32087 ERROR: mkFileDirR failed in mkDirForFilePath with status -520028
Jul 14 11:43:20 pid:32087 ERROR: Unable to make directory: "/irods-seq-sr04-ddn-gc10-33-34-35/jc18_testcol/testfile_263"
```

Obviously I can see that the resource has run out of space. However I am confused as to why the error messages are talking about making a directory as it should only be storing files, not creating directory's.

Also this resource has the high water mark set and it is in the structure shown below:
```
root -> replicate -> local  -> resource1
                            -> resource2 
                  -> remote -> resource3
                            -> resource4
```
local and remote are random resources. If resource1 is full I would expect the local resource to try and store the file in resource2. This does not appear to be happening. Have I understood the functionality of the high water mark properly?

**Comment by trel** [comment]
I think this is duplicate of https://github.com/irods/irods/issues/3195 - and we have a fix in CI.

Please compare and confirm.   Thanks.

**Comment by kellerb** [comment]
Which version or commit of irods are you using?

**Comment by kript** [comment]
We've just built packages for 4.1-stable ( 7bdac12a), and while this appears to have fixed #3215 (in dev, yet to deploy in production), I got similar errors with a similar for loop;

```
FileDirR :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- mes
sage [fileMkdir for [/irods-seq-sr04-ddn-gc10-33-34-35/jc18_testcol]]
        [-]     iRODS/server/drivers/src/fileDriver.cpp:290:fileMkdir :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [failed to call 'mkdir']
                [-]     libpassthru.cpp:354:passthru_file_mkdir_plugin :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [passthru_file_mkdir_plugin - failed calling child mkdir.]
                        [-]     librepl.cpp:863:replFileMkdir :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [replFileMkdir - Failed while calling child operation.]
                                [-]     librandom.cpp:412:random_file_mkdir :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [Failed calling child operation.]
                                        [-]     librandom.cpp:412:random_file_mkdir :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [Failed calling child operation.]
                                                [-]     libunixfilesystem.cpp:978:unix_file_mkdir_plugin :  status [UNIX_FILE_MKDIR_ERR]  errno [No space left on device] -- message [Mkdir error for "/irods-seq-sr04-ddn-gc10-33-34-35/jc18_testcol", errno = "No space left on device", status = -520028.]

Jul 14 16:20:51 pid:30440 ERROR: mkFileDirR failed in mkDirForFilePath with status -520028
Jul 14 16:20:51 pid:30440 ERROR: Unable to make directory: "/irods-seq-sr04-ddn-gc10-33-34-35/jc18_testcol/testcol2/testfile_230"
```

which is correct, but still not skipping;

```
iadmin lr irods-seq-sr04-ddn-gc10-33-34-35
resc_id: 12359
resc_name: irods-seq-sr04-ddn-gc10-33-34-35
zone_name: seq-dev
resc_type_name: unixfilesystem
resc_net: irods-g2-dev
resc_def_path: /irods-seq-sr04-ddn-gc10-33-34-35
free_space: 
free_space_ts Never
resc_info: 
r_comment: 
resc_status: 
create_ts 2016-06-14.11:18:19
modify_ts 2016-06-30.16:19:31
resc_children: 
resc_context: high_water_mark=12434804415
resc_parent: green9
resc_objcount: 5
irods-g2-dev@irods-g2-dev:~$ df -h /irods-seq-sr04-ddn-gc10-33-34-35
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop68     485M  480M  4.0K 100% /irods-seq-sr04-ddn-gc10-33-34-35
```

which implies the high_water_mark isn't being checked either?

**Comment by kellerb** [comment]
If I'm reading the output correctly, the `high_water_mark` is set to 12,434,804,415 bytes or about 11.5 GiB, and the filesystem has size 485 MiB. Because the `high_water_mark` is larger than the filesystem size, we wouldn't expect the `high_water_mark` to prevent iRODS creation operations on that resource.

If you set the `high_water_mark` to 500000000, the resource should be skipped as expected.

With regard to the directory being created: `unixfilesystem` resources don't create directories until they are asked to store files (i.e. `imkdir` creates a collection in iRODS, but does not create a directory in the vault). If the `iput` that failed was the first `iput` into that collection targeting the resource in question, that would explain why the directory was being created. When you look in the vault, does the directory exist?

Slightly related, once the fix for https://github.com/irods/irods/issues/3195 is pushed to 4-1-stable, this problem will also go away, because `unixfilesystem` resources will start performing free space checks in the voting phase (in addition to the `high_water_mark` checks they are already performing).

**Comment by kript** [comment]
My mistake, I thought the fix for #3195 had been pushed to stable; sorry - otherwise I wouldn't have bothered you. I *am* going to go away and check my maths however; that `high_water_mark` is indeed not right...

and yes, you're right of course; it must create the directory 'jc18_testcol' and if there is no free space...

So this is a duplicate, please close as such.
